binderhub/launcher.py

```diff
@@ -223,20 +241,54 @@ async def launch(self, image, username, server_name='', repo_url='', extra_args=
                 # Server hasn't actually started yet
-                # We wait for it!
-                # NOTE: This ends up being about ten minutes
-                for i in range(64):
-                    user_data = await self.get_user_data(escaped_username)
-                    if user_data['servers'][server_name]['ready']:
-                        break
-                    if not user_data['servers'][server_name]['pending']:
-                        raise web.HTTPError(500, "Image %s for user %s failed to launch" % (image, username))
-                    # FIXME: make this configurable
-                    # FIXME: Measure how long it takes for servers to start
-                    # and tune this appropriately
-                    await gen.sleep(min(1.4 ** i, 10))
+                # listen pending spawn (launch) events until server is ready
+                ready_event_container = []
+                buffer_list = []
+
+                async def handle_chunk(chunk):
+                    lines = b"".join(buffer_list + [chunk]).split(b"\n\n")
+                    # the last item in the list is usually an empty line ('')
+                    # but it can be the partial line after the last `\n\n`,
+                    # so put it back on the buffer to handle with the next chunk
+                    buffer_list[:] = [lines[-1]]
+                    for line in lines[:-1]:
+                        if line:
+                            line = line.decode("utf8", "replace")
+                        if line and line.startswith("data:"):
+                            event = json.loads(line.split(":", 1)[1])
+                            if event_callback:
+                                await event_callback(event)
+                            if event.get("ready", False):
+                                # stream ends when server is ready
+                                ready_event_container.append(event)
+
+                url_parts = ["users", username]
+                if server_name:
+                    url_parts.extend(["servers", server_name, "progress"])
                 else:
-                    raise web.HTTPError(500, f"Image {image} for user {username} took too long to launch")
+                    url_parts.extend(["server/progress"])
+                progress_api_url = url_path_join(*url_parts)
+                resp_future = self.api_request(
+                    progress_api_url,
+                    streaming_callback=handle_chunk,
+                    request_timeout=self.launch_timeout,
+                )
+                try:
+                    await gen.with_timeout(
+                        timedelta(seconds=self.launch_timeout), resp_future
+                    )
+                except (gen.TimeoutError, TimeoutError):
+                    raise web.HTTPError(
+                        500,
+                        "Image %s for user %s took too long to launch"
+                        % (image, username),
```

Comment:
Are you okay with f-strings? pyupgrade will upgrade some but not all. To make #1381 mergable in the future, I extracted out some manual changes regarding f-strings in #1387.

```suggestion
                        f"Image {image} for user {username} took too long to launch"
```